### PR TITLE
Directly use work.result() to retrieve tensor rather than passing as a separate argument

### LINKED ISSUE #43960: In DDP's Reducer, directly use work.result() to retrieve allreduced tensor.
## Refactor `forwardPassWorkHandle` to use `work.result()` in DDP's `Reducer`
We currently have a hack to retrieve an allreduced tensor from Python in C++ in `reducer.cpp`, where we store the resulting tensor in a struct: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L434

Now since https://github.com/pytorch/pytorch/pull/43386 is landed, we should refactor this code to simply pass the `work` object and retrieve the tensor with `result()`. 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski

### PR DESCRIPTION
**Pull request by bugra**
We currently are fetching an allreduced tensor from Python in C++ in, where we are storing the resulting tensor in a struct's parameter. This PR removes extra tensor paratemeter in the function parameter and fetch from a single place.

Fixes #43960

### CODE REVIEW

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.cpp:578]: I think `workHandle` is a pointer here

File: torch/nn/parallel/distributed.py
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:578]: We still need this tensor as its passed into `all_reduce` below.

**Review by rohan-varma — COMMENTED** [review]
Thank you for working on this! I left some comments and the PR should be good to go once those are resolved.

File: benchmarks/operator_benchmark/pt/qactivation_test.py
Comment (pritamdamania87) [inline benchmarks/operator_benchmark/pt/qactivation_test.py:53]: Are these changes intended to be part of this PR? Can we move these changes to a separate PR?

**Review by pritamdamania87 — COMMENTED** [review]
Could you update the PR title to accurately describe the changes in this PR?

File: benchmarks/operator_benchmark/pt/qactivation_test.py
Comment (bugra) [reply benchmarks/operator_benchmark/pt/qactivation_test.py]: @pritamdamania87 I moved to another PR.

File: torch/nn/parallel/distributed.py
Comment (bugra) [reply torch/nn/parallel/distributed.py]: Yep, my mistake. I kept it now.

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (bugra) [reply torch/csrc/distributed/c10d/reducer.cpp]: Yes, changed this as well.

**Comment by bugra** [comment]
@pritamdamania87  I updated both the description as well as the title of the PR. Let me know if this looks good.

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.cpp:578]: nit: Let's check `!results.empty()`

File: torch/csrc/distributed/c10d/reducer.h
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.h:92]: In addition to this, let's remove the tensor from `struct ForwardPassAllreduceWork`

**Review by rohan-varma — COMMENTED** [review]
Looks good! We should remove the now unneeded arg from `struct ForwardPassAllreduceWork` as I mentioned inline. After that, the diff should be good to go !

File: torch/csrc/distributed/c10d/reducer.h
Comment (bugra) [reply torch/csrc/distributed/c10d/reducer.h:92]: Removed

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (bugra) [reply torch/csrc/distributed/c10d/reducer.cpp]: Addressed

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks for working on this! Please wait for internal + OSS CI to pass before landing.

**Pull request closed without merging**
